Add spec and scripts for creating chocolatey package for crc

anjannath opened the pull request:
the package currently performs the setup steps like installing hyper-v creation of 'crc-users' group, adding current user to this group and the hyper-v administrators group. It doesn't install or setup the tray

it is a portable choco package and embedds the crc and crc-admin-helper-windows binaries

## How to test

Look at the contained README file for instructions on how to create the choco package: https://github.com/crc-org/crc/pull/3444/files#diff-34a4d7b03e9cd0af82b8172946c44f002f78f40e35bda3843d819cdbc64a8167

### To only test the installation of the chocolatey package use the following commands:
```posh
# install CRC from the temporary test feed
PS> choco install crc --version 2.12.0 --source https://www.myget.org/F/anjan-choco-test/api/v2 
PS> crc setup
PS> shutdown -r -t 00
PS> crc setup
PS> crc start # should successfully start the cluster
```

anjannath committed:
Add spec and scripts for creating chocolatey package for crc

the package currently performs the setup steps like installing hyper-v
creation of 'crc-users' group, adding current user to the
hyper-v administrators group. It doesn't install or setup the tray

it is a portable choco package and embedds the crc and crc-admin-helper-windows
binaries

anjannath committed:
choco: add github actions job to test building the chocolatey package

anjannath commented:
Have created a temporary test feed for easier testing, https://www.myget.org/feed/anjan-choco-test/package/nuget/crc. Use the following command to test the installation and see if everything is properly setup by the choco package:
```posh
PS> choco install crc --version 2.12.0 --source https://www.myget.org/F/anjan-choco-test/api/v2
PS> crc setup
PS> shutdown -r -t 00
PS> crc setup
PS> crc start # should successfully start the cluster
```

praveenkumar commented on pkg/crc/preflight/preflight_checks_windows.go:
```diff
@@ -119,12 +119,13 @@ func checkIfUserPartOfHyperVAdmins() error {
 
 func fixUserPartOfHyperVAdmins() error {
-	outGroupName, _, err := powershell.Execute(`(New-Object System.Security.Principal.SecurityIdentifier("S-1-5-32-578")).Translate([System.Security.Principal.NTAccount]).Value`)
+	_, _, err := powershell.ExecuteAsAdmin("adding current user to Hyper-V administrator group", fmt.Sprintf("Add-LocalGroupMember -SID 'S-1-5-32-578' -Member '%s'", username()))
```
Since we were executing this command without any admin privileges does this show  popup to user  now?

anjannath replied:
yes, UAC pop-up is shown, i think it'll be better to club the two "adding to crc-users" and "adding to Hyper-V admins" group together in a single check/fix function would be better, we'll get only one UAC pop-up then

praveenkumar replied:
But then does it also catch individual action error? iirc if we club 2-3 action in same then error catching is difficult.

anjannath replied:
No we can have combined stdout o/p from both the commands, but it'll not be able catch error from individual actions. i think in this case its an acceptable trade-off, the two actions are doing similar thing of adding current member to a group, so its easy to narrow down the issue

praveenkumar commented on pkg/crc/preflight/preflight_checks_windows.go:
```diff
@@ -126,8 +126,13 @@ func fixUserPartOfHyperVAdmins() error {
+func checkUserPartOfCrcUsers() error {
+	_, _, err := powershell.Execute(fmt.Sprintf("Get-LocalGroupMember -Group 'crc-users' -Member '%s'", username()))
+	return err
+}
```
we should mention in the commit logs about which commit removed that check and we are adding back.

anjannath replied:
ahh, i couldn't find the exact commit where it was removed, iirc it was during the time we were adding the msi

cfergeau replied:
As far as I can tell, this was https://github.com/crc-org/crc/commit/f98e79c41ea4f752204d2d70d4db705637f9f3d0
(found it using `git log -p pkg/crc/preflight/*win*.go` followed by a search for `Get-LocalGroupMember `. Can be done more directly using  `git log -p -G 'Get-LocalGroupMember'`)

praveenkumar commented on packaging/chocolatey/crc/crc.nuspec:
```diff
@@ -2,7 +2,7 @@
   <metadata>
     <id>crc</id>
-    <version>2.11.0</version>
+    <version>2.12.0</version>
```
do we need to do every single time when there is a release of crc?

anjannath replied:
yeah this needs to be updated for each release, running the `update.ps1` script will do this automatically

cfergeau replied:
You can introduce a `crc.nuspec.in` file with `__VERSION__` placeholders, this Makefile rule https://github.com/crc-org/crc/blob/76d07a985c9a7be95beba9b9c796760b6c258817/Makefile#L351-L356 will then autogenerate a `crc.nuspec` file with the right version in it when one runs `make packaging/chocolatey/crc/crc.nuspec`

anjannath replied:
wanted to try choco/windows specific tooling for updating the `nuspec` but this is better since we are already using similar method

cfergeau commented on pkg/crc/preflight/preflight_checks_windows.go:
```diff
@@ -126,8 +126,13 @@ func fixUserPartOfHyperVAdmins() error {
+	return err
+}
 
-	_, _, err = powershell.ExecuteAsAdmin("adding current user to Hyper-V administrator group", fmt.Sprintf("Add-LocalGroupMember -Group '%s' -Member '%s'", groupName, username()))
```
I believe this removal should be part of the previous commit, not of this one.

cfergeau commented on packaging/chocolatey/crc/crc.nuspec:
```diff
@@ -0,0 +1,37 @@
+    <licenseUrl>https://github.com/crc-org/crc/blob/main/LICENSE</licenseUrl>
+    <requireLicenseAcceptance>false</requireLicenseAcceptance>
+    <tags>crc ocp openshift openshiftlocal</tags>
+    <summary>CRC brings a minimal OpenShift Container Platform 4 cluster and Podman container runtime to your local computer</summary>
```
No 'podman' or 'okd' tag?

anjannath replied:
good point, will add those as well

cfergeau commented on packaging/chocolatey/crc/crc.nuspec:
```diff
@@ -0,0 +1,37 @@
+    <licenseUrl>https://github.com/crc-org/crc/blob/main/LICENSE</licenseUrl>
+    <requireLicenseAcceptance>false</requireLicenseAcceptance>
+    <tags>crc ocp openshift openshiftlocal</tags>
+    <summary>CRC brings a minimal OpenShift Container Platform 4 cluster and Podman container runtime to your local computer</summary>
```
Wondering if this should be 'Red Hat OpenShift Container Platform 4' rather than "OpenShift Container Platform 4"

cfergeau commented on packaging/chocolatey/crc/tools/chocolateyuninstall.ps1:
```diff
@@ -0,0 +1,11 @@
+Remove-LocalGroup -Name 'crc-users' -Confirm:$false | Out-Null
+
+# remove vsock registry element
+Remove-VsockGuestCommunicationRegistry | Out-Null
```
Isn't this similar to hyperv, we can't know if the vsock key was created by crc or not, so we should not remove it?

anjannath replied:
yeah right, also the check for existence of this needs to be improved currently its checking if the `ElementName` property exists, but we should check also the value of that property is `gvisor-tap-vsock`

anjannath replied:
not removing this registry key now

cfergeau replied:
ah, if the key has a very specific value, we can probably assume its ours.

cfergeau commented on packaging/chocolatey/crc/README.md:
````diff
@@ -0,0 +1,93 @@
+C:\Users\anath\redhat\crc\packaging\chocolatey\crc
+```
+3. Run the script `update.ps1` to update the package spec files with latest version numbers, the script fetches the latest tag from the Github repository and uses that as the latest version
+
````
Step 1. builds crc.exe from the current local git HEAD, why do we use the latest remote tag to version it? Shouldn't the version be derived from `git describe HEAD` instead?
If I do `git checkout v2.9.0; make out/windows-amd64/crc.exe` and then run `update.ps1`, it feels wrong to version this as `2.12.0`

anjannath replied:
this is handled now after addressing https://github.com/crc-org/crc/pull/3444#discussion_r1061604661 since everything is incorporated in the `Makefile`

cfergeau commented on packaging/chocolatey/crc/update.ps1:
```diff
@@ -0,0 +1,36 @@
+function global:au_BeforeUpdate {
+    # download admin helper release
+    $admin_helper_version = (Invoke-WebRequest https://api.github.com/repos/crc-org/admin-helper/tags | Select-Object -ExpandProperty Content | ConvertFrom-Json)[0].name
+    $Latest.URL64 = 'https://github.com/crc-org/admin-helper/releases/download/' + $admin_helper_version + '/crc-admin-helper-windows.exe'
```
Would it be possible to use `crc-embedder download` instead? This ensures we use the `crc-admin-helper` version expected by the currently checked out tree.

anjannath replied:
only issue is it downloads the `tray` zip along with the `admin-helper` binary, currently in the makefile, downloaded tray zip is removed after running `crc-embedder` and before running `choco pack` to avoid including it in the finale choco pkg. 

created https://github.com/crc-org/crc/pull/3475 to be able to select only the admin-helper to download

(we could also point crc-embedder to download to a temporary directory and we only copy the admin-helper binary to the tools/ directory)

anjannath committed:
Add user to crc-users and hyperv admins group in a single powershell admin call

this is done to avoid getting two UAC prompts during crc setup command

by combining both in a single call to powershell admin session we are losing
inidividual error messages to determine which action failed but in this case
its an acceptable trade-off, since the two actions are doing similar thing of
adding current member to a group, so its easy to narrow down the issue

anjannath committed:
simplify the checking for user part of the hyperv admins group

we don't need to convert the SID to a group name since the Add-LocalGroupMember
cmdlet has a -SID flag which can directly use the SID of the hyperv admin group
to add a member to it

anjannath committed:
add preflight check and fix to add the logon user to crc-users group

this is added back again as we are adding support for a chocolatey package
the problem this solves is that the chocolatey package scripts are ran  as
admin, and from an elevated powershell session we cannot easily capture the
logon username, to do this we need to start a powershell session without
administrator privileges from the eleveted powershell session but this
is not possible to do

it was removed in f98e79c41ea4f752204d2d70d4db705637f9f3d0

anjannath commented on packaging/chocolatey/crc/crc.nuspec.in:
```diff
@@ -0,0 +1,37 @@
+These runtimes provide minimal environments for development and testing purposes.
+CRC is mainly targeted at running on developers' desktops.
+For other OpenShift Container Platform use cases, such as headless or multi-developer setups, use the full [OpenShift installer](https://console.redhat.com/openshift/install).
+
```
Note: Need to mention about the pull-secret requirement here

cfergeau commented on pkg/crc/preflight/preflight_checks_windows.go:
```diff
@@ -99,40 +100,27 @@ func checkHyperVServiceRunning() error {
-func fixUserPartOfCrcUsers() error {
-	_, _, err := powershell.ExecuteAsAdmin("adding current user to crc-users group", fmt.Sprintf("Add-LocalGroupMember -Group 'crc-users' -Member '%s'", username()))
+func fixUserPartOfCrcUsersAndHypervAdminsGroup() error {
+	cmd := `Add-LocalGroupMember -Group 'crc-users' -Member '%s';Add-LocalGroupMember -SID 'S-1-5-32-578' -Member '%s'`
```
If the user is already part of `crc-users`, but not of  `S-1-5-32-578` does this work as expected? I mean, is `Add-LocalGroupMember -Group 'crc-users'` a no-op if we are already part of that group?

anjannath replied:
yes, its not a no-op it'll provide an error but since we are not checking the error reported by powershell, and powershell's default behaviour when an error occurs is to report and continue (`
$ErrorActionPreference=Continue`) it goes on to execute the next command

cfergeau commented on pkg/crc/preflight/preflight_checks_windows.go:
```diff
@@ -99,40 +100,27 @@ func checkHyperVServiceRunning() error {
 
-func checkIfUserPartOfHyperVAdmins() error {
-	// https://support.microsoft.com/en-us/help/243330/well-known-security-identifiers-in-windows-operating-systems
-	// BUILTIN\Hyper-V Administrators => S-1-5-32-578
```
I'd keep this comment, this seems to have been removed.

anjannath commented on packaging/chocolatey/crc/crc.nuspec.in:
```diff
@@ -0,0 +1,38 @@
+    <id>crc</id>
+    <version>__VERSION__</version>
+    <owners>https://github.com/crc-org</owners>
+    <title>CRC OpenShift 4</title>
```
@gbraad any suggestions on the title, it was initially `OpenShift Local` but then changed it to `CRC OpenShift 4`, should this be just `CRC` instead?

cfergeau replied:
`CRC` seems better than `CRC OpenShift 4`. If you want to refer to OpenShift, we should be using `Red Hat OpenShift Local`

anjannath replied:
updated to `CRC - Runs Containers` as we are using that in the README also

cfergeau replied:
I thought we had a consensus that CRC Runs Containers wasn't that great, but this is creeping back lately :-(

cfergeau commented on pkg/crc/preflight/preflight_checks_windows.go:
```diff
@@ -99,40 +100,30 @@ func checkHyperVServiceRunning() error {
-		logging.Debug(err.Error())
-		return fmt.Errorf("Failed checking if user is part of hyperv admins group")
+		logging.Debugf("current user is not part of 'crc-users' group: %v", err)
+		mErr.Collect(err)
```
Not clear why we want a multierror in this function? Looks like we could return as soon as we get an error?

anjannath replied:
yeah, its not needed.

cfergeau commented on pkg/crc/preflight/preflight_windows.go:
```diff
@@ -147,12 +125,26 @@ var adminHelperServiceCheks = []Check{
+	configKeySuffix:  "check-crc-users-group-exists",
+	checkDescription: "Checking if crc-users group exists",
+	check: func() error {
+		if _, _, err := powershell.Execute("Get-LocalGroup -Name crc-users"); err != nil {
```
Why this check? The group is supposed to be created by the msi installer/chocolatey, but we make sure this work has been done  before starting crc? Would this be caught indirectly by userPartOfCrcUsersAndHypervAdminsGroupCheck ?
I'd add a comment explaining why there is no 'fix' method (because it's done by chocolatey)

anjannath replied:
> Would this be caught indirectly by userPartOfCrcUsersAndHypervAdminsGroupCheck ?

yes, it'll be caught there as well, but in that case there's two possibilities either the user is not part of the group or the 'crc-users' group doesn't exits, this check existed from before and it runs before the userPartOfCrcUsersAndHypervAdminsGroupCheck 

apart from the choco or msi creating the group, this check is also _StartUpOnly_ so there's no fix for it, but now i think making it run for `crc setup` and adding a fix would be better -- for users who installed using msi, choco this'll be kinda no-op and those who did not will get an UAC and be able to proceed to using crc

cfergeau replied:
> apart from the choco or msi creating the group, this check is also StartUpOnly so there's no fix for it, but now i think making it run for crc setup and adding a fix would be better -- for users who installed using msi, choco this'll be kinda no-op and those who did not will get an UAC and be able to proceed to using crc

Makes sense indeed, but you can keep this for later.

cfergeau commented on Makefile:
```diff
@@ -378,3 +378,17 @@ $(BUILD_DIR)/windows-amd64/crc-windows-installer.zip: $(BUILD_DIR)/windows-amd64
+
+.PHONY: choco choco-clean
+CHOCO_PKG_DIR = packaging/chocolatey/crc
+$(CHOCO_PKG_DIR)/crc.nuspec: $(CHOCO_PKG_DIR)/crc.nuspec.in
```
I'm not sure this rule is needed.

anjannath replied:
yeah, removed, thanks

cfergeau commented on .github/workflows/windows-chocolatey.yml:
```diff
@@ -0,0 +1,31 @@
+  push:
+    branches:
+      - "main"
+      - "old-main"
```
we can drop `old-main`

cfergeau commented on packaging/chocolatey/crc/crc.nuspec.in:
```diff
@@ -0,0 +1,30 @@
+    <title>CRC - Runs Containers</title>
+    <authors>anjannath</authors>
+    <projectUrl>https://crc.dev/blog/about</projectUrl>
+    <iconUrl>https://i.ibb.co/m0b6dsL/crc-upstream-logo.png</iconUrl>
```
Forgot to ask before, but can we add this to some git repo, or to crc.dev, in short to some place we have control over?

anjannath replied:
this is supposed to be a CDN link to not rely on a single location, the current one is a random image hosting service i found online, but we can use more reliable free CDNs, the choco docs mentions some services but missed this in the PR

relevant choco docs link: https://docs.chocolatey.org/en-us/create/create-packages#package-icon-guidelines

cfergeau commented on packaging/chocolatey/crc/crc.nuspec.in:
```diff
@@ -0,0 +1,30 @@
+Learn more and contribute on [GitHub](https://github.com/crc-org/crc).
+    </description>
+    <releaseNotes>https://github.com/crc-org/crc/releases/tag/v__VERSION__</releaseNotes>
+    <copyright>2022 Red Hat, Inc</copyright>
```
This could become 2023 :)

cfergeau reviewed (approved):
A few minor comments left, but these can be addressed later.
